### PR TITLE
Document all of the directives

### LINKED ISSUE #867: Document all of the directives
It would be nice to have a documentation page with all of the directives.  @dleen has expressed interest in doing this.

### PR DESCRIPTION
**Pull request by dleen**
fixes #867

Need help with:

* `exec_doc`

I saw this was recently added in https://github.com/fastai/nbdev/issues/699 but I couldn't figure it out since I thought the cells were already executed...

* `nbflags`

It seems like this can be used to pass values to internal functions but I don't know of an example that we can use

* `default_cls_lvl`

It looks like this was supported in nbdev1, but is it implemented in 2?

* `hide_input`

This one sounded pretty easy but got confused because there is already a bunch of other ways to hide input so not sure what this is for.

@seeM @jph00 @hamelsmu

### CODE REVIEW

File: nbs/tutorial.ipynb [hunk 1]
```diff
@@ -18,9 +18,7 @@
    "source": [
     "# Tutorial\n",
     "\n",
-    "> A step-by-step guide to using nbdev\n",
-    "\n",
-    "- image: images/card.png"
```
Comment (hamelsmu) [inline nbs/tutorial.ipynb:23]: Why was this removed?

**Comment by hamelsmu** [comment]
@dleen I'm going to add some notes into #867 on where to find directives in the code (they are kind of spread out) and provide some pointers.  It's not necessarily that straightforward.  You've also asked some good questions which I'll try to address.  

I'll work on this morning, I will ping you when I'm done writing the notes.  

🙇🏽 Thanks for opening this PR

**Comment by hamelsmu** [comment]
@dleen I left some detailed comments [here](https://github.com/fastai/nbdev/issues/867#issuecomment-1216719986) to hopefully get you started, let me know if you have any questions

**Comment by hamelsmu** [comment]
Regarding the questions you've asked about those directives

- I believe `#|default_cls_lvl` is yet to be implemented in still on our TODO list
- I believe `#| nbflags` is kind of advanced at this point perhaps don't worry about this one just yet

The rest should be answered in the notes I added to the issue

File: nbs/tutorial.ipynb (hunk 1 above)
Comment (dleen) [reply nbs/tutorial.ipynb:23]: I noticed that it wasn't working: https://nbdev.fast.ai/tutorial.html and `card.png` is already used on the index page so I guessed that it wasn't intended here. What is this `-image: foo.png` syntax - it is a list? Is it a quarto extension to markdown?

**Comment by jph00** [comment]
`#|default_cls_lvl` is going to be moved into frontmatter and will change a bit anyway. `#| nbflags` is being removed.

Note that you can think of frontmatter as the same as directives, except directives are for a cell, whereas frontmatter is for the whole NB.

File: nbs/tutorial.ipynb (hunk 1 above)
Comment (jph00) [reply nbs/tutorial.ipynb:23]: The `>` thing certainly works - it adds the description to the metadata of the notebook and rendered page.

The bullet points with YAML underneath is an alternative way of adding frontmatter. I'm not sure it works yet (it used to in nbdev1) but it's something I'll be getting to in the next day or 2 if not.

Comment (dleen) [reply nbs/tutorial.ipynb:23]: Yes, I didn't delete the `>`. It's pushed to the next line in the git diff.

The bullet point thing definitely doesn't work. It's a bit confusing when things aren't doing anything (no errors, no output, although that should have been a clue in hindsight - that it didn't even output a string) and my bias is towards cleaning up... but if you plan to fix it I won't remove it here

Comment (hamelsmu) [reply nbs/tutorial.ipynb:23]: I think that bullet point thing used to work and then stopped working when we rolled back another PR supporting legacy front matter styles with markdown.  So good catch.

Seems like Jeremy is working on this right now though

Comment (dleen) [reply nbs/tutorial.ipynb:23]: Created an issue to track: https://github.com/fastai/nbdev/issues/873

File: nbs/directives.ipynb [hunk 2]
```diff
@@ -0,0 +1,828 @@
+{
```
Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <blockquote>that have are preceeded</blockquote><p> </p><p>I think should be "that are preceeded"</p><p> </p><blockquote>that do some kind of</blockquote><p> </p><p>Could be simplified to "and do"</p><p> </p><p>I would link to some python docs about the dunder all export. Most people I've talked to actually don't know about this, and this is a great opportunity to educate people by placing a link on that bullet point.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>This is a great explanation.  Perhaps briefly mention settings.ini as a way someone can check what the package is named <code>lib_name = ...</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>When we put <code>show_docs</code> in backticks here, this is a good opportunity to use a really cool nbdev feature you might not know about that does automated documentation linking. Instead of <code>show_docs</code> put the more qualified name <code>showdoc.show_doc</code> (no "s" at the end, BTW) and it should automatically link to the docs for that!</p><p> </p><p>Update: Jeremy added an entire "explanation" section for show_doc yesterday, which would be good to link to https://nbdev.fast.ai/explanations/showdoc.html</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>Great explanation.  You should also mention that naming your function or method with a <code>_</code> prefix does the same thing as well (also doesn't get placed in dunder all, not shown in docs etc)</p><p> </p><pre class="ql-syntax" spellcheck="false">#|export
def _private(): ...
</pre><p> </p><p>  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>s/show_docs/show_doc</p><p> </p><p>also "instead of showing docs" it is "in addition to show_doc, the source code is also displayed, for example..."</p><p> </p><p>This is because show_doc is still run and is showing the source code as well</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>"which results in:" -&gt; "which results in: (click to expand)"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>"When a cell does not have this annotation, the output is fixed ...." </p><p> </p><p>should be</p><p> </p><p>"When a cell does not have this annotation, it is run according to the default rules [described here](https://nbdev.fast.ai/explanations/showdoc.html)."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>Maybe provide an example to drive this home like a cell with "assert 1 == 2" and say when you run <code>nbdev_test_nbs</code> no error will occur.  </p><p> </p><p>Also mention that this may be useful for long running cells that you don't want to always test </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>Mention that this is equivalent to having these two quarto cell directives</p><p> </p><p>Allows you to hide the cell input</p><p>#| echo: false</p><p> </p><p>Allows you to hide the cell output</p><p>#| output: false</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>This is actually a kind of alias for #| notest  it does the same thing.  If you don't mind can you please verify this.  With this in mind you might want to combine the explanation of #| notest and #| eval: &lt;true|false&gt; </p><p> </p><p>nbdev respects both.  The latter is a quarto directive.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>I would hide the input cell with <code>#| echo: false</code> so that it renders nicely in the docs.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='57'/>

Comment (hamelsmu) [inline nbs/directives.ipynb:1]: <p>Read this tutorial that Jeremy authored yesterday https://nbdev.fast.ai/explanations/showdoc.html and perhaps link to it instead or update your explanation based on that.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/fastai/nbdev/pull/868/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='59'/>

**Comment by hamelsmu** [comment]
Thanks so much @dleen looking good.  I left some comments for you to consider.  Please let me know if you have any questions!

**Comment by hamelsmu** [comment]
@dleen updated some comments in case you are relying on email notifications (I would read them on GitHub instead).  Thanks again!

File: nbdev/directives.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# AUTOGENERATED! DO NOT EDIT! File to edit: ../nbs/directives.ipynb.
```
Comment (jph00) [inline nbdev/directives.py:1]: Remove this file

File: nbs/custom.yml [hunk 4]
```diff
@@ -8,6 +8,8 @@ website:
         href: getting_started.ipynb
       - text: "Tutorial"
         href: tutorial.ipynb
+      - text: "Directives"
```
Comment (jph00) [inline nbs/custom.yml:11]: I think just being in the sidebar is sufficient - no need to add here too

File: nbs/sidebar.yml [hunk 5]
```diff
@@ -3,6 +3,7 @@ website:
     contents:
       - getting_started.ipynb
       - tutorial.ipynb
+      - directives.ipynb
```
Comment (jph00) [inline nbs/sidebar.yml:6]: Suggest moving to the new "Explanations" section

File: nbs/directives.ipynb [hunk 6]
```diff
@@ -0,0 +1,828 @@
...
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "hide_input": false
+   },
+   "outputs": [],
+   "source": [
+    "#|default_exp directives"
```
Comment (jph00) [inline nbs/directives.ipynb:11]: Remove this - we're not exporting this to a python module

File: nbs/directives.ipynb [hunk 7]
```diff
@@ -0,0 +1,828 @@
...
+    "The documentation will look like this:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#|export\n",
```
Comment (jph00) [inline nbs/directives.ipynb:108]: Remove all `export` directives - we're not exporting a submodule.

File: nbs/directives.ipynb [hunk 8]
```diff
@@ -0,0 +1,828 @@
...
+    "for i in range(3):\n",
+    "    _secret() #|hide_line\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## `#|output: <true|false|asis>`\n",
```
Comment (jph00) [inline nbs/directives.ipynb:598]: This is handled by quarto, not nbdev, so I'd suggest removing these docs. Instead, you might want to have a section that mentions what quarto directives are, lists a couple of useful ones, and links to the quarto docs about them.

File: nbs/directives.ipynb [hunk 9]
```diff
@@ -0,0 +1,828 @@
...
+    "assert 1 + 1 == 2"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## `#|echo: <true|false>`\n",
+    "\n",
+    "This makes sure that only the output of a code cell is shown, not its input.\n",
```
Comment (jph00) [inline nbs/directives.ipynb:494]: This is a quarto directive - see comments below on `#|output`

File: nbs/directives.ipynb [hunk 10]
```diff
@@ -0,0 +1,828 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "Directives are special comments that have are preceeded by `#|` that do some kind of pre or post processing of a notebook data before docs are rendered. Some of these directives are part of Quarto, but others are unique to nbdev.\n",
+    "\n",
+    "See the [Quarto docs](https://quarto.org/docs/computations/execution-options.html) for even more execution options.\n",
```
Comment (jph00) [inline nbs/directives.ipynb:41]: Directives in nbdev and quarto aren't just for execution options. Perhaps link to a couple of other places Quarto documents directives, and mention the types of things that quarto directives can change.

I'd also suggest writing a bit here about the fact that both nbdev and quarto use directives, and that all quarto directives can be used in an nbdev notebook.

**Review by jph00 — CHANGES_REQUESTED** [review]
Many thanks for the PR @dleen . It looks great, and I've just got some minor requests for you.

It looks like you've accidentally exported a module from your tutorial notebook so we should fix that.

File: nbs/directives.ipynb (hunk 7 above)
Comment (hamelsmu) [reply nbs/directives.ipynb:108]: If you want to "mock" the effect of `#| export` for the docs you can manually call show_doc on it and hide the input.  like this:

```python
#|echo: false
show_doc(say_hello)
```

File: nbs/directives.ipynb (hunk 8 above)
Comment (hamelsmu) [reply nbs/directives.ipynb:598]: @dleen I talked to Jeremy about this.  I personally find this one-page cheat sheet of all the most important nbdev + quarto directives incredibly valuable.  This is what we settled on:

1. Have a section up front called "nbdev vs Quarto directives" that explains a bit about what directives are, and how Quarto has many of them, and nbdev adds additional ones.  Also mention that we will denote nbdev specific directives with an emoji like 📔 and quarto ones with an emoji like ❤️   .  Feel free to use different emojis if you are more creative than me!  
2. Categorize the directives as much as possible by "functionality" ex: (1) cell visibility (2) testing (3) exporting source code (4) etc ....
3. When you do write about quarto directives, do not write more than a couple of sentences about what they do, and link to the appropriate documentation in quarto in https://quarto.org/ (LMK if you need help finding any links) just because we don't want to duplicate the quarto docs too much

File: nbs/directives.ipynb (hunk 10 above)
Comment (hamelsmu) [reply nbs/directives.ipynb:41]: See comment: https://github.com/fastai/nbdev/pull/868#discussion_r948232459

File: nbs/directives.ipynb (hunk 9 above)
Comment (hamelsmu) [reply nbs/directives.ipynb:494]: See comment: https://github.com/fastai/nbdev/pull/868#discussion_r948232459

File: nbs/directives.ipynb (hunk 8 above)
Comment (dleen) [reply nbs/directives.ipynb]: Tried my best here but probably didn't delete enough. See https://dleen.github.io/nbdev/explanations/directives.html

**Comment by dleen** [comment]
View the latest version: https://dleen.github.io/nbdev/explanations/directives.html

**Comment by hamelsmu** [comment]
@dleen thanks so much for these changes!  I'm going to do some very minor cleanup things on this branch before merging if that is ok.

**Comment by dleen** [comment]
Yep, no objections! At this point it's probably far quicker to just make the edits than type them out here as comments

**Comment by hamelsmu** [comment]
@dleen can you add me as a collaborator on your fork please?

**Comment by dleen** [comment]
> ~@dleen can you add me as a collaborator on your fork please?~ Nevermind
> 
> (I am still working on it will continue tomorrow)

Done!

**Comment by hamelsmu** [comment]
Ok, I actually made some significant changes to the layout of this, I tried really hard to maximize the readability of this cheat sheet.  

- Changed the introduction section to give more of an overview of directives, and introduce the emojis
- Categorized the directives by functionality
- Placed examples of using directives in collapsible callouts. Without this visual distinction, the document started to become too visually confusing and unnavigable.
- Placed a header on the cell execution section of the show_doc explainer page so that I could link to it from this directive page.

**Comment by Isaac-Flath** [comment]
This is awesome!  This will definitely be a go-to resource for me :)

One additional directive I use often with quarto is `#| include: false` which combines `echo` and `output` to hide both code and output that may be nice to include in that section.

**Comment by Isaac-Flath** [comment]
I am just realizing the quarto's `#| hide` does the same thing as Quarto's `#| include` so maybe not needed after all if nbdev's implementation is preferred.

**Comment by hamelsmu** [comment]
Amazing find @Isaac-Flath where did you find that in the Quarto docs?   Perhaps we should deprecate `#|hide` in favor of `#|include` ....

**Comment by jph00** [comment]
I'm not sure. I can see it's a bit weird having two ways to do the same thing. OTOH I use `#|hide` so frequently, it's nice to have it be so short.
